File: bundles/org.eclipse.vjet.vsf.vjo.kernel/src/org/eclipse/vjet/vsf/RemoteReqtHdl.java
/*******************************************************************************
 * Copyright (c) 2012 eBay Inc. and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     eBay Inc. - initial API and implementation
 *******************************************************************************/
package org.eclipse.vjet.vsf;

import org.eclipse.vjet.dsf.javatojs.anno.AJsProxy;
import org.eclipse.vjet.dsf.dap.proxy.NativeJsProxy;
import org.mozilla.mod.javascript.Scriptable;
import org.eclipse.vjet.dsf.javatojs.anno.AExclude;
import org.eclipse.vjet.dsf.dap.proxy.INativeJsFuncProxy;
import org.eclipse.vjet.dsf.dap.proxy.NativeJsFuncProxy;
import org.eclipse.vjet.dsf.dap.proxy.NativeJsTypeRef;
import org.eclipse.vjet.dsf.javatojs.anno.AJavaOnly;
import org.eclipse.vjet.vsf.ITransport;
import org.eclipse.vjet.vsf.Message;

//NativeJsProxy for vjo.dsf.RemoteReqtHdl.js
@org.eclipse.vjet.dsf.resource.utils.CodeGen("NativeJsProxyGenerator")
@AJsProxy
public class RemoteReqtHdl extends NativeJsProxy implements ITransport {

    /** for framework use only */
    @AExclude
    public RemoteReqtHdl(Scriptable nativeObj){
        super(nativeObj);
    }

    /** internal use only */
    protected RemoteReqtHdl(Object ...args){
        super(args);
    }

    public RemoteReqtHdl() {
        super();
    }

    public void handleRequest(Message message, Object callback) {
        callWithName("handleRequest", message, callback);
    }

    public void callback(Message message) {
        callWithName("callback", message);
    }

    @AJavaOnly
    public static final NativeJsTypeRef<RemoteReqtHdl> prototype = NativeJsTypeRef.get(RemoteReqtHdl.class);

    public final INativeJsFuncProxy<RemoteReqtHdl> handleRequest = NativeJsFuncProxy.create(this, "handleRequest");

    public final INativeJsFuncProxy<RemoteReqtHdl> callback = NativeJsFuncProxy.create(this, "callback");
}

File: vsf/bundles/org.eclipse.vjet.vsf.vjo.kernel/src/org/eclipse/vjet/vsf/RemoteReqtHdl.js
/*******************************************************************************
 * Copyright (c) 2012 eBay Inc. and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     eBay Inc. - initial API and implementation
 *******************************************************************************/
vjo.ctype('org.eclipse.vjet.vsf.RemoteReqtHdl') //< public
.satisfies('org.eclipse.vjet.vsf.ITransport')
.needs(['org.eclipse.vjet.vsf.SvcErr',
'org.eclipse.vjet.vsf.ServiceRegistry',
'org.eclipse.vjet.vsf.ServiceResponse',
'org.eclipse.vjet.vsf.ServiceEngine'
        ])
//> needs org.eclipse.vjet.vsf.IBinding, org.eclipse.vjet.vsf.ITransport, org.eclipse.vjet.vsf.Message
.protos({
	//> public void handleRequest(Message message, Object callback)
	handleRequest : function(message, callback) {
//		message.trace += '-->RemoteHdl_' + message.svcId;
		message.response = new org.eclipse.vjet.vsf.ServiceResponse(); //< ServiceResponse
		// svcConfig is required
		if (!message.svcConfig) {
	   		org.eclipse.vjet.vsf.SvcErr.err(message, org.eclipse.vjet.vsf.SvcErr.InvRq, "message.svcConfig is undefined")
			return;
		}
		// get binding
		var binding = org.eclipse.vjet.vsf.ServiceRegistry.getBinding(message.svcConfig.reqtMarshalling); //< IBinding
		if (binding){
			binding.serialize(message);
		}
		else {
			// no binding error
			org.eclipse.vjet.vsf.SvcErr.err(message, org.eclipse.vjet.vsf.SvcErr.InvRqBnd, "reqtMarshalling=" + message.svcConfig.reqtMarshalling);
		}
		this.invoke(message);	
	},
	//> private void invoke(Message message);
	invoke : function(message) {
		// use transport if registered then return
		// JSCALLBACK uses JSONP transport
		var tT = ((message.svcConfig.respMarshalling == 'JSCALLBACK')?'JSONP':'XHR');
		var t = org.eclipse.vjet.vsf.ServiceRegistry.getTransport(tT); //< ITransport
		if (t) {
			t.handleRequest(message, this.callback);
		}
	},
	//> public void callback(Message message);
	callback: function (message) {
		try {
			var binding = org.eclipse.vjet.vsf.ServiceRegistry.getBinding(message.svcConfig.respMarshalling); //< IBinding
			if (binding){
				binding.deserialize(message);
			}
			else {
				org.eclipse.vjet.vsf.SvcErr.err(message, org.eclipse.vjet.vsf.SvcErr.InvRspBnd, "marshalling=" + message.svcConfig.respMarshalling + ":responseText=" + message.response);
			}
		} catch (e) {
			org.eclipse.vjet.vsf.SvcErr.err(message, org.eclipse.vjet.vsf.SvcErr.InvRsp, "marshalling=" + message.svcConfig.respMarshalling + ":responseText=" + message.response);
		}
		// TODO review if commented out then DapSample works
		org.eclipse.vjet.vsf.ServiceEngine.handleResponse(message);
	}
})
.endType();